ComponentKit/HostingView/CKComponentHostingView.mm

```diff
@@ -126,49 +128,114 @@ - (void)setContext:(id<NSObject>)context
 - (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
                             rootIdentifier:(CKComponentScopeRootIdentifier)rootIdentifier
                      didReceiveStateUpdate:(id (^)(id))stateUpdate
-                     tryAsynchronousUpdate:(BOOL)tryAsynchronousUpdate
+                                      mode:(CKUpdateMode)mode
 {
   CKAssertMainThread();
   _pendingStateUpdates.insert({globalIdentifier, stateUpdate});
-  [self _setComponentNeedsUpdate];
+  [self _setNeedsUpdateWithMode:mode];
 }
 
 #pragma mark - Private
 
-- (void)_setComponentNeedsUpdate
+- (void)_setNeedsUpdateWithMode:(CKUpdateMode)mode
 {
-  if (!_componentNeedsUpdate) { // Avoid thrashing delegate
-    _componentNeedsUpdate = YES;
-    [self setNeedsLayout];
-    [_delegate componentHostingViewDidInvalidateSize:self];
+  if (_componentNeedsUpdate && _requestedUpdateMode == CKUpdateModeSynchronous) {
+    return; // Already scheduled a synchronous update; nothing more to do.
+  }
+
+  _componentNeedsUpdate = YES;
+  _requestedUpdateMode = mode;
+
+  switch (mode) {
+    case CKUpdateModeAsynchronous:
+      [self _asynchronouslyUpdateComponentIfNeeded];
+      break;
+    case CKUpdateModeSynchronous:
+      [self setNeedsLayout];
+      [_delegate componentHostingViewDidInvalidateSize:self];
+      break;
+  }
+}
+
+- (void)_asynchronouslyUpdateComponentIfNeeded
+{
+  if (_scheduledAsynchronousComponentUpdate) {
+    return;
+  }
+  _scheduledAsynchronousComponentUpdate = YES;
+
+  // Wait until the end of the run loop so that if multiple async updates are triggered we don't thrash.
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self _scheduleAsynchronousUpdate];
+  });
+}
+
+- (void)_scheduleAsynchronousUpdate
+{
+  if (_requestedUpdateMode != CKUpdateModeAsynchronous) {
+    // A synchronous update was either scheduled or completed, so we can skip the async update.
+    _scheduledAsynchronousComponentUpdate = NO;
+    return;
   }
+
+  id<NSObject> modelCopy = _currentModel;
+  id<NSObject> contextCopy = _currentContext;
+  CKComponentScopeRoot *scopeRootCopy = _scopeRoot;
+  auto stateUpdatesToApply = std::make_shared<const CKComponentStateUpdateMap>(_pendingStateUpdates);
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    const CKBuildComponentResult result = CKBuildComponent(scopeRootCopy, *stateUpdatesToApply, ^CKComponent *{
+      return [_componentProvider componentForModel:modelCopy context:contextCopy];
+    });
+    dispatch_async(dispatch_get_main_queue(), ^{
+      if (!_componentNeedsUpdate) {
+        // A synchronous update snuck in and took care of it for us.
+        return;
+      }
+
+      if (_currentModel == modelCopy
```

Comment:
I'd had a bit more comments regarding the fact that state updates will run concurrently and hence we could get out of order execution and this is why we need this check.